keps/3125-maximum-execution-time/README.md

```diff
@@ -0,0 +1,207 @@
+
+### Goals
+
+Introduce a Job agnostic API of setting and enforcing the maximum execution time of a Job.
+
+
+### Non-Goals
+
+Change the behavior of the Kueue's scheduler based on the maximum execution time.
+
+## Proposal
+
+A new Kueue specific label is added `kueue.x-k8s.io/max-exec-time-seconds` containing
+the maximum number of seconds the Job is expected to execute.
+
+The duration is passed to the Job's Workload.
+
+If a workload stays admitted for longer that it is expected it will be Deactivated. 
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Batch User, I want to have the ability to set a limit on how long my job can run.
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+## Design Details
+
+### API
+
+#### Workload Spec
+
+Add a new optional filed to hold the 
+
+```go
+// WorkloadSpec defines the desired state of Workload
+// +kubebuilder:validation:XValidation:rule="has(self.priorityClassName) ? has(self.priority) : true", message="priority should not be nil when priorityClassName is set"
+type WorkloadSpec struct {
+	// ...
+
+	// MaximumExecutionTime if provided, determines the maximum time the workload can be admitted  
+	// befrore it's automatically deactivated.
+	MaximumExecutionTime *metav1.Duration `json:"maximumExecutionTime,omitempty"`
+}
+```
+
+#### Workload Status
+
+Add a optional filed to hold the 
+```go
+// WorkloadStatus defines the observed state of Workload
+type WorkloadStatus struct {
+	// ...
+
+	// AccumulatedExecutionTime holds the total duration the workload spent in Admitted state
+	// in the previous `Admit` - `Evict` cycles.
+	AccumulatedExecutionTime metav1.Duration `json:"maximumExecutionTime,omitempty"`
+}
+
+```
+
+### Constants
+
+Define a new constant holding the new label name `kueue.x-k8s.io/max-exec-time-seconds`.
+
+### Controller
+
+#### Workload
+During reconcile:
+
+- When a workload transitions out of `Admitted` state, add the time spent in its `status.AccumulatedExecutionTime`
+```go
+cond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadAdmitted)
+wl.Status.AccumulatedExecutionTime += wl.Spec.MaximumExecutionTime.Duration -
+    time.Now().Sub(cond.LastTransitionTime.Time)
+```
+- When the workload is admitted and a maximum execution time is specified compute the remaining execution execution time as:
+```go
+if wl.Spec.MaximumExecutionTime != nil {
+	if cond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadAdmitted); cond != nil && cond.Status == metav1.ConditionTrue {
+		remainingTime := wl.Spec.MaximumExecutionTime.Duration -
+            wl.Status.AccumulatedExecutionTime -
+            time.Now().Sub(cond.LastTransitionTime.Time)
+	}
+}
+```
+  - If `remainingTime` > 0 , a reconcile request should be queued with a `remainingTime` delay.
+  - If `remainingTime` <= 0 , the workload should be deactivated `wl.spec.Active = *false`.
```

Comment:
Will there be an explicit message/event/condition explaining why the workload was deactivated?

Reply:
sure, and we should also reset the accumulated time, otherwise reactivating the wl will be ineffective